File: go/circular_queue.go
package structures

import (
	"errors"
	"fmt"
)

type CircularQueue struct {
	capacity, frontIndex, tailIndex, size uint64
	items                                 []*int64
}

func NewCircularQueue(capacity uint64) *CircularQueue {
	return &CircularQueue{
		capacity:   capacity,
		frontIndex: 0,
		tailIndex:  0,
		size:       0,
		items:      []*int64{},
	}
}

func (cq *CircularQueue) clear() {
	var i = cq.frontIndex

	for i != cq.tailIndex {
		cq.items[i] = nil

		i = (func() uint64 {
			if cq.isLast(i) {
				return 0
			}

			return i + 1
		})()
	}
}

func (cq *CircularQueue) dequeue() (*int64, error) {
	if cq.empty() {
		return nil, errors.New("[circular-queue] is empty!")
	}

	var item = cq.items[cq.frontIndex]

	cq.frontIndex = (func() uint64 {
		if cq.isLast(cq.frontIndex) {
			return 0
		}

		return cq.frontIndex + 1
	})()
	cq.size--

	return item, nil
}

func (cq *CircularQueue) enqueue(value *int64) (*int64, error) {
	if cq.full() {
		return nil, errors.New("[circular-queue] is full!")
	}

	cq.tailIndex = (func() uint64 {
		if cq.isLast(cq.tailIndex) {
			return 0
		}

		return cq.tailIndex - 1
	})()
	cq.items[cq.tailIndex] = value

	return value, nil
}

func (cq *CircularQueue) front() *int64 {
	return cq.items[cq.frontIndex]
}

func (cq *CircularQueue) tail() *int64 {
	return cq.items[cq.tailIndex]
}

func (cq *CircularQueue) show() {
	var i = cq.frontIndex

	for i != cq.tailIndex {
		fmt.Println(cq.items[i])

		i = (func() uint64 {
			if cq.isLast(i) {
				return 0
			}

			return i + 1
		})()
	}
}

func (cq *CircularQueue) isLast(index uint64) bool {
	return index == cq.capacity-1
}

func (cq *CircularQueue) full() bool {
	return cq.size == cq.capacity
}

func (cq *CircularQueue) empty() bool {
	return cq.size == 0
}

File: rust/src/circular_queue.rs
use std::io::{Error, ErrorKind};

struct CircularQueue {
    capacity: u64,
    front_index: u64,
    tail_index: u64,
    size: u64,
    items: Vec<Option<i64>>,
}

impl CircularQueue {
    pub fn new(capacity: u64) -> Self {
        Self {
            capacity,
            front_index: 0,
            tail_index: 0,
            size: 0,
            items: Vec::with_capacity(capacity as usize),
        }
    }

    pub fn clear(&mut self) {
        let mut i = self.front_index;

        while i != self.tail_index {
            self.items[i as usize] = None;
            i = self.nextIndex(i);
        }
    }

    pub fn dequeue(&mut self) {
        if self.empty() {
            return;
        }

        self.front_index = self.nextIndex(self.front_index);
        self.size -= 1;
    }

    pub fn enqueue(&mut self, item: i64) -> Result<(), Error> {
        if self.full() {
            return Err(Error::new(ErrorKind::Other, "[queue] is full!"));
        }

        self.tail_index = self.nextIndex(self.tail_index);
        self.items[self.tail_index as usize] = Some(item);
        self.size += 1;

        Ok(())
    }

    pub fn front(&self) -> Option<i64> {
        self.items[self.front_index as usize]
    }

    pub fn tail(&self) -> Option<i64> {
        self.items[self.tail_index as usize]
    }

    pub fn show(&self) {
        let mut i = self.front_index;

        while i != self.tail_index {
            println!("{}", self.items[i as usize].unwrap());
            i = self.nextIndex(i);
        }
    }

    pub fn nextIndex(&self, index: u64) -> u64 {
        if index == self.capacity - 1 {
            return 0;
        }

        index + 1
    }

    pub fn full(&self) -> bool {
        self.size == self.capacity
    }

    pub fn empty(&self) -> bool {
        self.size == 0
    }
}
